File: Excel/1000-1099/1028/Challenge 1028.py
import pandas as pd

path = "1000-1099/1028/1028 Extract Name and Score.xlsx"
input = pd.read_excel(path, usecols="A", nrows=20, skiprows=1)
test = pd.read_excel(path, usecols="B:C", nrows=20, skiprows=1)

pattern = r'^\d+,("[^"]*"|\'[^\']*\'|[^,]*),(?:"[^"]*"|\'[^\']*\'|[^,]*),(?:"[^"]*"|\'[^\']*\'|[^,]*),("[0-9]+"|\'[0-9]+\'|[0-9]+)$'

result = (
    input["CSV Data"]
    .str.extract(pattern)
    .set_axis(["Name", "Score"], axis=1)
    .assign(
        Name=lambda x: x["Name"]
        .str.strip("'\"")
        .apply(lambda s: " ".join(reversed(s.split(", "))) if "," in s else s),
        Score=lambda x: x["Score"].str.strip("'\"").astype(int),
    )
)

print(result.equals(test))
# True

File: Excel/1000-1099/1028/Challenge 1028.R
library(tidyverse)
library(readxl)

path <- "1000-1099/1028/1028 Extract Name and Score.xlsx"
input <- read_excel(path, range = "A2:A21")
test <- read_excel(path, range = "B2:C21")

pattern <- paste0(
  "^(\\d+),",
  "(\"[^\"]*\"|'[^']*'|[^,]*),",
  "(\"[^\"]*\"|'[^']*'|[^,]*),",
  "(\"[^\"]*\"|'[^']*'|[^,]*),",
  "(\"\\d+\"|'\\d+'|\\d+)$"
)

result <- input %>%
  extract(
    `CSV Data`,
    into = c("id", "name", "department", "field", "score"),
    regex = pattern
  ) %>%
  mutate(
    across(name:score, \(x) str_remove_all(x, "^['\"]|['\"]$")),
    name = if_else(
      str_detect(name, ","),
      str_replace(name, "^(.+),\\s*(.+)$", "\\2 \\1"),
      name
    ),
    score = as.numeric(score)
  ) %>%
  select(Name = name, Score = score)

all.equal(result, test)
# True
